hail/src/main/scala/is/hail/expr/types/physical/PSubsetStruct.scala

```diff
@@ -0,0 +1,215 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{Region, UnsafeUtils}
+import is.hail.asm4s.{Code, Settable, SettableBuilder, Value, const}
+import is.hail.expr.ir.{EmitCodeBuilder, EmitMethodBuilder, IEmitCode}
+import is.hail.expr.types.BaseStruct
+import is.hail.expr.types.virtual.TStruct
+import is.hail.utils._
+
+object PSubsetStruct {
+  def apply(ps: PStruct, fieldNames: String*): PSubsetStruct = {
+    val f = fieldNames.toArray
+    PSubsetStruct(ps, f)
+  }
+}
+
+// Semantics: PSubsetStruct is a non-constructible view of another PStruct, which is not allowed to mutate
+// that underlying PStruct's region data
+final case class PSubsetStruct(ps: PStruct, _fieldNames: Array[String]) extends PStruct {
+  override lazy val fieldNames = _fieldNames
```

Comment:
I kind of hate this

Reply:
why do you need to override?

Reply:
Because it's defined on PBaseStruct

Reply:
And with a lazy modifier, which means I can't just override it in the argument-list portion of the constructor